crates/bevy_ecs/src/query/iter.rs

```diff
@@ -157,6 +159,640 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
+            .transmute_filtered::<(L, Entity), F>(world.components());
+
+        // SAFETY:
+        // `self.world` has permission to access the required components.
+        // The original query iter has not been iterated on, so no items are aliased from it.
+        let query_lens = unsafe {
+            query_lens_state.iter_unchecked_manual(
+                world,
+                world.last_change_tick(),
+                world.change_tick(),
+            )
+        };
+        let mut keyed_query: Vec<_> = query_lens
+            .map(|(key, entity)| (key, NeutralOrd(entity)))
+            .collect();
+        keyed_query.sort_unstable();
+        let entity_iter = keyed_query.into_iter().map(|(.., entity)| entity.0);
+        // SAFETY:
+        // `self.world` has permission to access the required components.
+        // Each lens query item is dropped before the respective actual query item is accessed.
+        unsafe {
+            QuerySortedIter::new(
+                world,
+                self.query_state,
+                entity_iter,
+                world.last_change_tick(),
+                world.change_tick(),
+            )
+        }
+    }
+
+    /// Sorts all query items into a new iterator with a comparator function over the query lens.
+    ///
+    /// This sort is stable (i.e., does not reorder equal elements).
+    ///
+    /// This uses [`slice::sort_by`] internally.
+    ///
+    /// Defining the lens works like [`transmute_lens`](crate::system::Query::transmute_lens).
+    /// This includes the allowed parameter type changes listed under [allowed transmutes].
+    /// However, the lens uses the filter of the original query when present.
+    ///
+    /// [allowed transmutes]: crate::system::Query#allowed-transmutes
+    ///
+    /// # Panics
+    ///
+    /// This will panic if `next` has been called on `QueryIter` before, unless the underlying `Query` is empty.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// # use std::ops::Deref;
+    /// #
+    /// # impl Deref for PartValue {
+    /// #     type Target = f32;
+    /// #
+    /// #     fn deref(&self) -> &Self::Target {
+    /// #         &self.0
+    /// #     }
+    /// # }
+    /// #
+    /// # let mut world = World::new();
+    /// #
+    /// #[derive(Component)]
+    /// struct PartValue(f32);
+    ///
+    /// // We can use a cmp function on components do not implement Ord.
+    /// fn system_1(query: Query<&PartValue>) {
+    ///     // Sort part values according to `f32::total_comp`.
+    ///     let part_values: Vec<&PartValue> = query
+    ///         .iter()
+    ///         .sort_by::<&PartValue, _>(|value_1, value_2| value_1.total_cmp(*value_2))
+    ///         .collect();
+    /// }
+    /// #
+    /// # let mut schedule = Schedule::default();
+    /// # schedule.add_systems((system_1));
+    /// # schedule.run(&mut world);
+    /// ```
+    pub fn sort_by<L: ReadOnlyQueryData + 'w, Fn>(
```

Comment:
You can get rid of the generics you need to elide at call site here and elsewhere by using `compare: impl FnMut(..)` instead.

Reply:
Sadly that requires two unstable Rust features, I list those under Future Possibilities above.

Reply:
I don't get any compile errors when I clone the fork and change it locally :thinking:

Reply:
hmm, I can try again, what I can think of is that I might only have tried on the sort_by_key methods, which have both the Fn trait and its returned key as a generic.
When attempting this, I did get two separate errors specifically for impl Fn traits and `impl Trait` returned by Fn traits though. 🤔